src/qibo/parallel.py

```diff
@@ -0,0 +1,193 @@
+                cls, *args, **kwargs)
+        return cls._instance
+
+    def run(self, params=None):
+        """Evaluates the custom_function object for a specific set of
+        parameters. This class performs the lock mechanism to duplicate objects
+        for each process.
+        """
+        # lock to avoid race conditions
+        self.lock.acquire()
+        # get process name
+        pname = self.mp.current_process().name
+        # check if there are objects already stored
+        args = self._objects_per_process.get(pname, None)
+        if args is None:
+            args = []
+            for obj in self.arguments:
+                try:
+                    # copy object if copy method is available
+                    copy = obj.copy(deep=True)
+                except AttributeError:
+                    # if copy is not implemented just use the original object
+                    copy = obj
+                except Exception as e:
+                    # use print otherwise the message will not appear
+                    print('Exception in ParallelResources', str(e))
+                args.append(copy)
+            args = tuple(args)
+            self._objects_per_process[pname] = args
+        # unlock
+        self.lock.release()
+        # finally compute the custom function
+        return self.custom_function(params, *args)
+
+    def reset(self):
+        """Cleanup memory."""
+        self._objects_per_process = {}
+        self.custom_function = None
+        self.lock = None
+        self.arguments = ()
+
+
+def _executor(params): # pragma: no cover
+    """Executes singleton call."""
+    return ParallelResources().run(params)
+
+
+def parallel_execution(circuit, states, processes=None):
+    """Execute circuit for multiple states.
+
+    Example:
+        ::
+
+            from qibo import models, set_threads
+            from qibo.parallel import parallel_execution
+            import numpy as np
+            # create circuit
+            nqubits = 22
+            circuit = models.QFT(nqubits)
+            # create random states
+            states = [ np.random.random(2**nqubits) for i in range(5)]
+            # set threads to 1 per process (optional, requires tuning)
+            set_threads(1)
+            # execute in parallel
+            results = parallel_execution(circuit, states, processes=2)
+
+    Args:
+        circuit (qibo.models.Circuit): the input circuit.
+        states (list): list of states for the circuit evaluation.
+        processes (int): number of processes for parallel evaluation.
+
+    Returns:
+        Circuit evaluation for input states.
+    """
+    if states is None or not isinstance(states, list):  # pragma: no cover
+        from qibo.config import raise_error
+        raise_error(RuntimeError, "states must be a list.")
+
+    _check_parallel_configuration(processes)
+
+    def operation(state, circuit): # pragma: no cover
+        return circuit(state)
```

Comment:
Do you think it would be useful and simple to add an `nshots` parameter here when calling the circuit? This will allow the user to simulate measurements, because currently both parallelization methods can only be used to get the final state vector. I guess for this to work `nshots` should also be added as parameter in the `parallel_execution` function.